src/storage/src/source/mysql/replication.rs

```diff
@@ -110,84 +118,113 @@ pub(crate) fn render<G: Scope<Timestamp = TransactionId>>(
+                assert!(prev < *part, "duplicate or unsorted partitions");
+
+                // Create a partition representing all the UUIDs in the gap between the previous one and this one
+                let gap = GtidPartition::new_range(prev, part.before(), TransactionId::Absent);
+                future_partitions.insert(gap);
+
+                // Set prev to the UUID + 1 so the next gap starts after this one
+                prev = part.after();
+
+                // Mint capabilities for this partition at the resume transaction-id
+                let part_ts = GtidPartition::new_singleton(part.clone(), ts.timestamp().clone());
+                trace!(%id, "timely-{worker_id} creating partition capability for {part_ts:?}");
+                let cap = GtidPartitionCapability::new(&part_ts, data_cap_set, upper_cap_set);
+                active_partition_caps.insert(part.clone(), cap);
+            }
+            future_partitions.insert(GtidPartition::new_range(prev, Uuid::maximum(), TransactionId::Absent));
+
+            data_cap_set.downgrade(&*future_partitions);
+            upper_cap_set.downgrade(&*future_partitions);
+            trace!(%id, "timely-{worker_id} replication reader started at {:?}", resume_upper);
 
             let mut rewinds = BTreeMap::new();
             while let Some(event) = rewind_input.next().await {
                 if let AsyncEvent::Data(caps, data) = event {
                     for req in data {
-                        let snapshot_transaction_id = TransactionId::new(req.snapshot_gtid_set.first().expect("at least one gtid").latest_transaction_id());
-                        if resume_transaction_id > snapshot_transaction_id + TransactionId::new(1) {
-                            let err = DefiniteError::BinlogCompactedPastResumePoint(resume_transaction_id.to_string(), (snapshot_transaction_id + TransactionId::new(1)).to_string());
-                            // If the replication stream cannot be obtained from the resume point there is
-                            // nothing else to do. These errors are not retractable.
-                            for (output_index, _) in table_info.values() {
-                                let update = ((*output_index, Err(err.clone())), TransactionId::MAX, 1);
-                                data_output.give(&data_cap_set[0], update).await;
+                        let mut rewind_partition_map = BTreeMap::new();
+                        for snapshot_partition in req.snapshot_gtid_set {
+                            let part = snapshot_partition.interval().singleton().unwrap();
+                            let snapshot_time = snapshot_partition.timestamp();
+
+                            // Check that the replication stream can be resumed from the snapshot point or before.
+                            if !binlog_start_gtids.get(part).is_some_and(|binlog_start_gtid| binlog_start_gtid.timestamp() <= &(*snapshot_time + TransactionId::Active(1))) {
```

Comment:
Assuming `snapshot_gtid_set` becomes an `Antichain<GtidSet>` as suggested above (and presumably renamed to `snapshot_upper`) this check here is simply:

```rust
if !PartialOrder::less_equal(&resume_upper, &snapshot_upper) {
    // definite error
}
```